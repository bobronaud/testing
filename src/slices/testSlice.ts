import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../app/store';

export type Question = {
  type: string;
  question: string;
  answers: string[] | null;
  rightAnswer: string | string[] | null;
};

type Answer = Question & {
  userAnswer: string;
};

type Phase = 'starting' | 'inProcess' | 'finished';

type Completed = {
  [key: number]: boolean;
};

interface state {
  phase: Phase;
  questions: Question[];
  currentQuestion: Question | null;
  currentStep: number;
  completedSteps: Completed;
  result: Answer[];
}
const initialState: state = {
  phase: 'starting',
  questions: [],
  currentQuestion: null,
  currentStep: 0,
  completedSteps: {},
  result: [],
};

const testSlice = createSlice({
  name: 'test',
  initialState,
  reducers: {
    writeAnswer: (state, action: PayloadAction<string | string[]>) => {
      const userAnswer = action.payload;
      const answer = Object.assign(state.currentQuestion as Question, {
        userAnswer,
      }) as Answer;
      state.result = [...state.result, answer];
    },
    finishStep: (state) => {
      const isLastStep = state.currentStep + 1 === state.questions.length;
      if (isLastStep) {
        state.phase = 'finished';
      } else {
        state.completedSteps[state.currentStep] = true;
        state.currentStep += 1;
        state.currentQuestion = state.questions[state.currentStep];
      }
    },
    setPhase: (state, action: PayloadAction<Phase>) => {
      const newPhase = action.payload;
      state.phase = newPhase;
    },
    uploadQuestions: (state, action: PayloadAction<Question[]>) => {
      const questions = action.payload;
      state.questions = questions;
      state.currentQuestion = questions[0];
    },
  },
});

export default testSlice.reducer;
export const selectCards = (state: RootState) => state.test;
export const { writeAnswer, finishStep, setPhase, uploadQuestions } = testSlice.actions;
