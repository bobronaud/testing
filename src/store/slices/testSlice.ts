import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';

export type Question = {
  type: string;
  question: string;
  answers: string[] | null;
  rightAnswer: string | string[] | null;
};

type Answer = Question & {
  userAnswer: string;
};

export type Phase = 'starting' | 'inProcess' | 'finished';

type Completed = {
  [key: number]: boolean;
};

export type State = {
  timer: boolean;
  time: [number, number];
  phase: Phase | null;
  questions: Question[];
  currentQuestion: Question | null;
  currentStep: number;
  completedSteps: Completed;
  result: Answer[];
};
const initialState: State = {
  timer: true,
  time: [5, 0],
  phase: null,
  questions: [],
  currentQuestion: null,
  currentStep: 0,
  completedSteps: {},
  result: [],
};

const testSlice = createSlice({
  name: 'test',
  initialState,
  reducers: {
    writeAnswer: (state, action: PayloadAction<string | string[]>) => {
      const userAnswer = action.payload;
      const answer = Object.assign(state.currentQuestion as Question, {
        userAnswer,
      }) as Answer;
      state.result = [...state.result, answer];
    },
    finishStep: (state) => {
      const isLastStep = state.currentStep + 1 === state.questions.length;
      if (isLastStep) {
        state.phase = 'finished';
      } else {
        state.completedSteps[state.currentStep] = true;
        state.currentStep += 1;
        state.currentQuestion = state.questions[state.currentStep];
      }
    },
    setPhase: (state, action: PayloadAction<Phase>) => {
      const newPhase = action.payload;
      state.phase = newPhase;
    },
    setTimer: (state, action: PayloadAction<boolean>) => {
      const value = action.payload;
      state.timer = value;
    },
    setTime: (state, action: PayloadAction<[number, number]>) => {
      const time = action.payload;
      state.time = time;
    },
    uploadQuestions: (state, action: PayloadAction<Question[]>) => {
      const questions = action.payload;
      state.questions = questions;
      state.currentQuestion = questions[0];
    },
    setDataApp: (state, action: PayloadAction<State>) => {
      const data = action.payload;
      const newState = Object.assign(state, data);
      state = newState;
    },
    resetProgress: (state) => {
      const newState = Object.assign(state, initialState);
      state = newState;
    },
  },
});

export default testSlice.reducer;
export const selectCards = (state: RootState) => state.test;
export const {
  writeAnswer,
  finishStep,
  setPhase,
  setTimer,
  setTime,
  setDataApp,
  uploadQuestions,
  resetProgress,
} = testSlice.actions;
