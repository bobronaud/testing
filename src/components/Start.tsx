import { Box, Button, FormControlLabel, Switch, Typography } from '@mui/material';

import type { Question } from '../slices/testSlice';
import { setPhase, uploadQuestions } from '../slices/testSlice';
import { useAppDispatch } from '../app/hooks';
import data from '../assets/test.json';

const Start = () => {
  const dispatch = useAppDispatch();
  const handleclick = () => {
    dispatch(setPhase('inProcess'));
    dispatch(uploadQuestions(data as Question[]));
  };
  return (
    <Box
      sx={{
        width: '50%',
        height: '50%',
        boxShadow: '0 0 3px 3px rgba(0, 0, 0, 0.2)',
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center',
        gap: 3,
      }}>
      <Typography variant='h4' component='h1'>
        Тестирование
      </Typography>
      <FormControlLabel
        control={<Switch defaultChecked />}
        label='Ограничение по времени: 5 мин'
      />
      <Button variant='contained' onClick={handleclick}>
        Начать
      </Button>
    </Box>
  );
};

export default Start;
