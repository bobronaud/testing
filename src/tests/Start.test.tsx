import userEvent from '@testing-library/user-event';
import Start from '../components/Start';
import { State } from '../store/slices/testSlice';
import renderWithRedux from './helpers/renderWithRedux';
import { chooseMany, chooseOne } from './fixtures/questionsData';
import { getByTestId } from '@testing-library/react';

describe('тест компонента Start', () => {
  const initialState = {
    test: {
      timer: true,
      time: [5, 0],
      phase: 'starting',
      questions: [],
      currentQuestion: null,
      currentStep: 0,
      completedSteps: {},
      result: [],
    } as State,
  };
  test('отрисовка', () => {
    const container = renderWithRedux(<Start />, initialState);
    expect(container).toMatchSnapshot();
  });

  test('работоспособность чекера', async () => {
    const { getByRole } = renderWithRedux(<Start />, initialState);
    const cheker = getByRole('checkbox');
    expect(cheker).toBeChecked();
    await userEvent.click(cheker);
    expect(cheker).not.toBeChecked();
  });
});
