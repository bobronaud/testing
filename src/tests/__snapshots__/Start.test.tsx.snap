// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`тест компонента Start отрисовка 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-1txp0i9"
      >
        <h1
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          Тестирование
        </h1>
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-byenzh-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Ограничение по времени: 5мин
          </span>
        </label>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Начать
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1txp0i9"
    >
      <h1
        class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      >
        Тестирование
      </h1>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
        >
          <span
            class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-byenzh-MuiButtonBase-root-MuiSwitch-switchBase"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Ограничение по времени: 5мин
        </span>
      </label>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Начать
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
